Poor diagnostics on missing semicolon after match statement

I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8a78e534c893d563ba3bf81786161178)):

```rust
fn stuff() -> i32 {
    match 1i32 {
        2i32 => 3i32,
        _ => panic!("wrong"),
    }
    4i32
}
```

I expected to see this happen:
A compiler error about missing `;` between `}` and `4i32`

Instead, this happened:
```rust
error[E0308]: `match` arms have incompatible types
 --> src/lib.rs:4:14
  |
2 | /     match 1i32 {
3 | |         2i32 => 3i32,
  | |                 ---- this is found to be of type `i32`
4 | |         _ => panic!("wrong"),
  | |              ^^^^^^^^^^^^^^^ expected `i32`, found `()`
5 | |     }
  | |_____- `match` arms have incompatible types
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

This is very confusing, because I "know" that `panic!()` return type is `!` which is coercible to anything else (or whatever is the technical explanation for "we can really ignore the return type of `panic!`"), so it should not ever be present in error messages about mistyping stuff.

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.44.0-beta.3 (0f0d70055 2020-05-11)
binary: rustc
commit-hash: 0f0d70055530bdbf3f0acb7b8ad25aa4a6ad8ea9
commit-date: 2020-05-11
host: x86_64-pc-windows-msvc
release: 1.44.0-beta.3
LLVM version: 9.0
```

This issue is also present in 1.34.2 (from a year ago), and in two-week-old nightly (rustc 1.45.0-nightly (99cb9ccb9 2020-05-11))


p.s.: When deminimizing the issue, i realized that it's even more confusing when match is the last statement in a for loop (as it was in my real code), i.e.:
```rust
fn stuff2(){
    for _ in 0..5{
        match 1i32 {
            2i32 => 3i32,
            _ => panic!("wrong"),
        }
    }
}
```
gives the same error message about incompatible match arms.

--- a/compiler/rustc_hir_typeck/src/_match.rs
+++ b/compiler/rustc_hir_typeck/src/_match.rs
@@ -121,6 +121,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         prior_arm_ty,
                         prior_arm_span,
                         scrut_span: scrut.span,
+                        scrut_hir_id: scrut.hir_id,
                         source: match_src,
                         prior_arms: other_arms.clone(),
                         opt_suggest_box_span,

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -775,6 +775,7 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                 ref prior_arms,
                 opt_suggest_box_span,
                 scrut_span,
+                scrut_hir_id,
                 ..
             }) => match source {
                 hir::MatchSource::TryDesugar(scrut_hir_id) => {
@@ -841,6 +842,18 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                         arm_span,
                     ) {
                         err.subdiagnostic(subdiag);
+                    }
+                    if let Some(hir::Node::Expr(m)) = self.tcx.hir().find_parent(scrut_hir_id)
+                        && let Some(hir::Node::Stmt(stmt)) = self.tcx.hir().find_parent(m.hir_id)
+                        && let hir::StmtKind::Expr(_) = stmt.kind
+                    {
+                        err.span_suggestion_verbose(
+                            stmt.span.shrink_to_hi(),
+                            "consider using a semicolon here, but this will discard any values \
+                             in the match arms",
+                            ";",
+                            Applicability::MaybeIncorrect,
+                        );
                     }
                     if let Some(ret_sp) = opt_suggest_box_span {
                         // Get return type span and point to it.

--- a/compiler/rustc_middle/src/traits/mod.rs
+++ b/compiler/rustc_middle/src/traits/mod.rs
@@ -541,6 +541,7 @@ pub struct MatchExpressionArmCause<'tcx> {
     pub prior_arm_ty: Ty<'tcx>,
     pub prior_arm_span: Span,
     pub scrut_span: Span,
+    pub scrut_hir_id: hir::HirId,
     pub source: hir::MatchSource,
     pub prior_arms: Vec<Span>,
     pub opt_suggest_box_span: Option<Span>,

--- a/tests/ui/suggestions/issue-81839.stderr
+++ b/tests/ui/suggestions/issue-81839.stderr
@@ -4,15 +4,22 @@ error[E0308]: `match` arms have incompatible types
 LL | /     match num {
 LL | |         1 => {
 LL | |             cx.answer_str("hi");
-   | |             --------------------
-   | |             |                  |
-   | |             |                  help: consider removing this semicolon
-   | |             this is found to be of type `()`
+   | |             -------------------- this is found to be of type `()`
 LL | |         }
 LL | |         _ => cx.answer_str("hi"),
    | |              ^^^^^^^^^^^^^^^^^^^ expected `()`, found future
 LL | |     }
    | |_____- `match` arms have incompatible types
+   |
+help: consider removing this semicolon
+   |
+LL -             cx.answer_str("hi");
+LL +             cx.answer_str("hi")
+   |
+help: consider using a semicolon here, but this will discard any values in the match arms
+   |
+LL |     };
+   |      +
 
 error: aborting due to previous error
 

--- a/tests/ui/wf/wf-unsafe-trait-obj-match.stderr
+++ b/tests/ui/wf/wf-unsafe-trait-obj-match.stderr
@@ -11,6 +11,10 @@ LL | |     }
    |
    = note: expected reference `&S`
               found reference `&R`
+help: consider using a semicolon here, but this will discard any values in the match arms
+   |
+LL |     };
+   |      +
 
 error[E0038]: the trait `Trait` cannot be made into an object
   --> $DIR/wf-unsafe-trait-obj-match.rs:26:21

